### PR TITLE
handle async error

### PR DESCRIPTION
**Pull request by TingDaoK**
*Issue #, if available:*

- To handle Asynchronous Error from S3 https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html#S3AsynchronousErrorCodeList

*Description of changes:*

- For CompleteMultipartUpload, that mentioned with async error code, which is the only API we called from GET/PUT/COPY that will ever has async error, we parse the body and get the error code if there.
- If error is InternalErrors, we do retry on the request. Other errors fall into non-retryable Error.

TODO: Don't have tests for the corner case. 

Note: The default type may receive async error as well. CRT S3 client **DO NOT** handle those, and never will parse the body for default type of meta request. It's the caller's responsibility to handle it properly. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/s3/private/s3_util.h [hunk 1]
```diff
@@ -172,11 +181,21 @@ AWS_S3_API
...
 /* Get a top-level (exists directly under the root tag) tag value. */
+AWS_S3_API
 struct aws_string *get_top_level_xml_tag_value(
     struct aws_allocator *allocator,
     const struct aws_byte_cursor *tag_name,
     struct aws_byte_cursor *xml_body);
 
+/* Get a top-level (exists directly under the root tag) tag value with expected root name. */
+AWS_S3_API
+struct aws_string *get_top_level_xml_tag_value_with_root_name(
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply include/aws/s3/private/s3_util.h:192]: hmmm, it's a non-public helper function. I feel either way is kind awkward from the use case... 

It serves the use case of getting the error code from the body if exist. If the function is as you suggested, we cannot tell the difference between an real failure or the body is a success response that don't have the tag.

File: source/s3_meta_request.c [hunk 2]
```diff
@@ -1070,11 +1070,37 @@ static void s_s3_meta_request_send_request_finish(
...
+            aws_reset_error();
+            return AWS_ERROR_SUCCESS;
+        } else {
+            return aws_last_error();
+        }
+    } else {
+        /* Check the error code. Map the S3 error code to CRT error code. */
+        int error_code = aws_s3_crt_error_code_from_server_error_code_string(error_code_string);
+        if (error_code != AWS_ERROR_S3_INTERNAL_ERROR) {
+            /* All error besides of internal error from async error doesn't recoverable from retry for now. */
```
Comment (DmitriyMusatkin) [inline source/s3_meta_request.c:1091]: nit: doesn't -> are not

**Review by DmitriyMusatkin — APPROVED** [review]

File: include/aws/s3/private/s3_util.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (DmitriyMusatkin) [reply include/aws/s3/private/s3_util.h]: we should really come up with a better way to test helper functions without exposing them in the library interface. 

Anyways, im fine with it with the new naming
